### ISSUE TITLE
Regression on Out_channel exceptions

### ISSUE THREAD

**Issue by jmid** [ocaml/ocaml#12898]
As part of revising our `Out_channel` `multicoretests` I discovered the following regression.
Consider the following program writing to a closed channel twice:
```ocaml
let path = "/tmp/foo.txt" in
let oc = Out_channel.open_text path in
let () = Out_channel.close oc in
let () =
  try Out_channel.output_char oc '1'; Printf.printf " 1 OK%!"
  with Sys_error _ -> Printf.printf " 1 Sys_error%!" in
let () =
  try Out_channel.output_char oc '2'; Printf.printf " 2 OK%!"
  with Sys_error _ -> Printf.printf " 2 Sys_error%!" in
()
```

On 4.14.1, 5.0.0, 5.1.0, 5.1.1 this behaves the same, yielding an exception in both cases:
```
$ ocaml /tmp/ocbug.ml 
 1 Sys_error 2 Sys_error
```

However, on `5.2` and on `trunk/5.3` this no longer raises an exception in the second case.
```
$ ocaml /tmp/ocbug.ml 
 1 Sys_error 2 OK
```

The issue is not limited to `output_char`. The other output operations `output_byte`, `output_string`, ... are also affected.

Looking at changes in `runtime/io.c` I've established that the behaviour changed with #12314.
A compiler built from the previous commit to `trunk` / `5.2` exhibits the previous behaviour:
- c717b6b47e - (6 months ago) `Merge pull request #12114 from fabbing/tsan_patch`

**Comment by jmid** [comment]
Looking some more into this, AFAICS the following assignment added in #12314
https://github.com/ocaml/ocaml/blob/1680bfe231c062e91c383610953c4559b18f0102/runtime/io.c#L252-L255
is at odds with this later comment explaining how `caml_flush_partial` is used to trigger a `Sys_error`
https://github.com/ocaml/ocaml/blob/1680bfe231c062e91c383610953c4559b18f0102/runtime/io.c#L706-L709

For the above case of `Out_channel.output_char`, after the assignment the following `Putch` condition to trigger `caml_flush_partial` no longer holds, which explains why the behaviour changes at the second `Out_channel.output_char`
https://github.com/ocaml/ocaml/blob/1680bfe231c062e91c383610953c4559b18f0102/runtime/io.c#L281-L284

**Comment by jmid** [comment]
Polite ping to @xavierleroy as the author of #12314 (but likely busy preparing an interesting course)

**Comment by gasche** [comment]
Indeed, it may be wise to plan for another maintainer to take this on for the 5.2 milestone. I might have a look if no one else is volunteering.

**Comment by jmid** [comment]
Incidentally, today I discovered that this bug was independently discovered and fixed in https://github.com/ocaml/ocaml/pull/12678

**Comment by xavierleroy** [comment]
The fix is probably easy, and might be "pickable" from #12678, but I don't know when I can look into this.  So either someone else works on a fix for 5.2, or we revert #12314 from 5.2 and work on this issue as part of #12678.

**Comment by gasche** [comment]
I submitted a proposal to fix this in #12947.
